coding.go

```diff
@@ -34,64 +34,63 @@ func (nb _PBNode__NodeBuilder) DecodeDagProto(r io.Reader) error {
 	if err := pbn.Unmarshal(encoded); err != nil {
 		return fmt.Errorf("unmarshal failed. %v", err)
 	}
-	pbLinks := make([]PBLink, 0, len(pbn.Links))
-	for _, link := range pbn.Links {
-		hash, err := cid.Cast(link.GetHash())
+	return fluent.Recover(func() {
+		fb := fluent.WrapAssembler(nb)
+		fb.CreateMap(0, func(fmb fluent.MapAssembler) {
+			fmb.AssembleEntry("Links").CreateList(len(pbn.Links), func(flb fluent.ListAssembler) {
+				for _, link := range pbn.Links {
+					hash, err := cid.Cast(link.GetHash())
 
-		if err != nil {
-			return fmt.Errorf("unmarshal failed. %v", err)
-		}
-		pbLinks = append(pbLinks, PBLink{
-			d: PBLink__Content{
-				Hash: MaybeLink{
-					Maybe: schema.Maybe_Value,
-					Value: Link{cidlink.Link{Cid: hash}},
-				},
-				Name: MaybeString{
-					Maybe: schema.Maybe_Value,
-					Value: String{link.GetName()},
-				},
-				Tsize: MaybeInt{
-					Maybe: schema.Maybe_Value,
-					Value: Int{int(link.GetTsize())},
-				},
-			},
+					if err != nil {
+						panic(fluent.Error{Err: fmt.Errorf("unmarshal failed. %v", err)})
+					}
+					flb.AssembleValue().CreateMap(0, func(fmb fluent.MapAssembler) {
+						fmb.AssembleEntry("Hash").AssignLink(cidlink.Link{Cid: hash})
+						fmb.AssembleEntry("Name").AssignString(link.GetName())
+						fmb.AssembleEntry("Tsize").AssignInt(int(link.GetTsize()))
+					})
+				}
+			})
+			fmb.AssembleEntry("Data").AssignBytes(pbn.GetData())
 		})
-	}
-	nb.nd.d.Links.x = pbLinks
-	nb.nd.d.Data.x = pbn.GetData()
-	return nil
+	})
 }
 
 // EncodeDagProto is a fast path encoding to protobuf
 // for PBNode types
 func (nd PBNode) EncodeDagProto(w io.Writer) error {
 	pbn := new(merkledag_pb.PBNode)
-	pbn.Links = make([]*merkledag_pb.PBLink, 0, len(nd.d.Links.x))
-	for _, link := range nd.d.Links.x {
+	pbn.Links = make([]*merkledag_pb.PBLink, 0, nd.FieldLinks().Length())
+	linksIter := nd.FieldLinks().ListIterator()
+	for !linksIter.Done() {
+		_, nlink, err := linksIter.Next()
+		if err != nil {
+			return err
+		}
+		link := nlink.(PBLink)
 		var hash []byte
-		if link.d.Hash.Maybe == schema.Maybe_Value {
-			cid := link.d.Hash.Value.x.(cidlink.Link).Cid
+		if link.FieldHash().Exists() {
+			cid := link.FieldHash().Must().Link().(cidlink.Link).Cid
 			if cid.Defined() {
 				hash = cid.Bytes()
 			}
 		}
 		var name *string
-		if link.d.Name.Maybe == schema.Maybe_Value {
-			tmp := link.d.Name.Value.x
+		if link.FieldName().Exists() {
+			tmp := link.FieldName().Must().String()
```

Comment:
I'd write a review about how all this causes unfortunate allocations, but... it's unavoidable in interacting with protobufs, as I understand it.  Ah well.

(Fun fact if we ever come back to optimize this though, without just tearing deep through and doing the parse protobuf ourselves completely: I think we could bump these `tmp` things into a struct and make sure the whole thing escapes to heap at once, and it'll shave a constant factor off.)